src/core/machine-id-setup.c

```diff
@@ -180,68 +187,99 @@ int machine_id_setup(const char *root, sd_id128_t machine_id, sd_id128_t *ret) {
+                r = detach_mount_namespace();
+                if (r < 0)
+                        return log_error_errno(r, "Failed to set up new mount namespace: %m");
 
-        if (umount(etc_machine_id) < 0)
-                return log_error_errno(errno, "Failed to unmount transient %s file in our private namespace: %m", etc_machine_id);
+                if (umount(etc_machine_id) < 0)
+                        return log_error_errno(errno, "Failed to unmount transient %s file in our private namespace: %m", etc_machine_id);
 
-        /* Update a persistent version of etc_machine_id */
-        r = id128_write(etc_machine_id, ID128_PLAIN, id, true);
-        if (r < 0)
-                return log_error_errno(r, "Cannot write %s. This is mandatory to get a persistent machine ID: %m", etc_machine_id);
+                /* Update a persistent version of etc_machine_id */
+                r = id128_write(etc_machine_id, ID128_PLAIN, id, true);
+                if (r < 0)
+                        return log_error_errno(r, "Cannot write %s. This is mandatory to get a persistent machine ID: %m", etc_machine_id);
 
-        /* Return to initial namespace and proceed a lazy tmpfs unmount */
-        r = namespace_enter(-1, initial_mntns_fd, -1, -1, -1);
-        if (r < 0)
-                return log_warning_errno(r, "Failed to switch back to initial mount namespace: %m.\nWe'll keep transient %s file until next reboot.", etc_machine_id);
+                /* Return to initial namespace and proceed a lazy tmpfs unmount */
+                r = namespace_enter(-1, initial_mntns_fd, -1, -1, -1);
+                if (r < 0)
+                        return log_warning_errno(r, "Failed to switch back to initial mount namespace: %m.\nWe'll keep transient %s file until next reboot.", etc_machine_id);
 
-        if (umount2(etc_machine_id, MNT_DETACH) < 0)
-                return log_warning_errno(errno, "Failed to unmount transient %s file: %m.\nWe keep that mount until next reboot.", etc_machine_id);
+                if (umount2(etc_machine_id, MNT_DETACH) < 0)
+                        return log_warning_errno(errno, "Failed to unmount transient %s file: %m.\nWe keep that mount until next reboot.", etc_machine_id);
 
-        return 0;
+                return 0;
+        } else {
+                /* File does not exist or is not a mountpoint
+                 *
+                 * We read the existing machine-id  in /etc */
+                fd = open(etc_machine_id, O_RDONLY|O_CLOEXEC|O_NOCTTY);
+                if (fd > 0) {
+                        r = id128_read_fd(fd, ID128_PLAIN, &id);
+                        fd = safe_close(fd);
+                        if (r == 0)
+                                /* /etc/machine-id already contains a valid machine-id, don't do anything */
+                                return 0;
+                }
+                if (fd < 0 && errno != ENOENT)
```

Comment:
Same issue here: `errno` can only be used directly after a glibc operation that is documented to return errno. Here, the call to id128_read_fd() "destroys" the errno value, in the sense that it might be changed.